internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher_test.go

```diff
@@ -6,74 +6,1558 @@ package githubupstreamwatcher
-				Spec: v1alpha1.GitHubIdentityProviderSpec{
-					GitHubAPI: v1alpha1.GitHubAPIConfig{
-						Host: ptr.To("127.0.0.1"),
-						TLS: &v1alpha1.TLSSpec{
-							CertificateAuthorityData: "irrelevant",
+			name: "no GitHubIdentityProviders",
+		},
+		{
+			name:    "happy path with all fields",
+			secrets: []runtime.Object{goodSecret},
+			githubIdentityProviders: []runtime.Object{
+				validFilledOutIDP,
+			},
+			wantResultingCache: []*upstreamgithub.ProviderConfig{
+				{
+					Name:               "some-idp-name",
+					ResourceUID:        "some-resource-uid",
+					Host:               fmt.Sprintf("https://%s", *validFilledOutIDP.Spec.GitHubAPI.Host),
+					UsernameAttribute:  "id",
+					GroupNameAttribute: "name",
+					OAuth2Config: &oauth2.Config{
+						ClientID:     "some-client-id",
+						ClientSecret: "some-client-secret",
+					},
+					AllowedOrganizations:    []string{"organization1", "org2"},
+					OrganizationLoginPolicy: "OnlyUsersFromAllowedOrganizations",
+					AuthorizationURL:        fmt.Sprintf("https://%s/login/oauth/authorize", *validFilledOutIDP.Spec.GitHubAPI.Host),
+					HttpClient:              buildPretendHttpClient(t, goodServer),
+				},
+			},
+			wantResultingUpstreams: []v1alpha1.GitHubIdentityProvider{
+				{
+					ObjectMeta: validFilledOutIDP.ObjectMeta,
+					Spec:       validFilledOutIDP.Spec,
+					Status: v1alpha1.GitHubIdentityProviderStatus{
+						Phase: v1alpha1.GitHubPhaseReady,
+						Conditions: []metav1.Condition{
+							buildClientCredentialsObtainedTrue(t, validFilledOutIDP.Spec.Client.SecretName),
+							buildGitHubConnectionValidTrue(t, *validFilledOutIDP.Spec.GitHubAPI.Host),
+							buildHostValidTrue(t, *validFilledOutIDP.Spec.GitHubAPI.Host),
+							buildOrganizationsPolicyValidTrue(t, *validFilledOutIDP.Spec.AllowAuthentication.Organizations.Policy),
+							buildTLSConfigurationValidTrue(t),
+						},
+					},
+				},
+			},
+			wantLogs: []string{
+				buildLogForUpdatingClientCredentialsObtained("some-idp-name", "True", "Success", fmt.Sprintf(`clientID and clientSecret have been read from spec.client.SecretName (\"%s\")`, validFilledOutIDP.Spec.Client.SecretName)),
+				buildLogForUpdatingOrganizationPolicyValid("some-idp-name", "True", "Success", fmt.Sprintf(`spec.allowAuthentication.organizations.policy (\"%s\") is valid`, string(*validFilledOutIDP.Spec.AllowAuthentication.Organizations.Policy))),
+				buildLogForUpdatingHostValid("some-idp-name", "True", "Success", `spec.githubAPI.host (\"%s\") is valid`, *validFilledOutIDP.Spec.GitHubAPI.Host),
+				buildLogForUpdatingTLSConfigurationValid("some-idp-name", "True", "Success", "spec.githubAPI.tls.certificateAuthorityData is valid"),
+				buildLogForUpdatingGitHubConnectionValid("some-idp-name", "True", "Success", `spec.githubAPI.host (\"%s\") is reachable and TLS verification succeeds`, *validFilledOutIDP.Spec.GitHubAPI.Host),
+				buildLogForUpdatingPhase("some-idp-name", "Ready"),
+			},
+		},
+		{
```

Comment:
Do we have cases that cover these situations:

```json
{
  name: "...loop preserves existing unchanged status conditions...",
}, {
  name: "....loop only updates relevant timestamps and leaves unchanged status timestamps as-is...",
}
```

Reply:
added